src/consensus/extendedmetadata.cpp

```diff
@@ -0,0 +1,153 @@
+// Copyright (c) 2022 Logos Foundation
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#include <blockindex.h>
+#include <consensus/activation.h>
+#include <consensus/merkle.h>
+#include <consensus/params.h>
+#include <consensus/validation.h>
+#include <hash.h>
+#include <logging.h>
+#include <streams.h>
+#include <validation.h>
+#include <version.h>
+
+#include <consensus/extendedmetadata.h>
+
+static const size_t EPOCH_MERKLE_ROOT_LENGTH = 33;
+
+std::vector<uint256> CollectEpochMerkleLeaves(const CBlock &block,
+                                              const CBlockIndex *pindexPrev) {
+    const uint256 currentEpochBlockHash = block.hashEpochBlock;
+    std::vector<uint256> epochMerkleLeaves;
+    size_t dummyNumLayers;
+
+    // Add block's own merkle root
+    // 0 for the block hash, as we cannot know our own block hash here
+    epochMerkleLeaves.push_back(
+        ComputeMerkleRoot({uint256(), block.hashMerkleRoot}, dummyNumLayers));
+
+    // Walk back chain until epoch hash changes
+    while (pindexPrev && pindexPrev->hashEpochBlock == currentEpochBlockHash) {
+        // Add H(block hash || merkle root)
+        epochMerkleLeaves.push_back(ComputeMerkleRoot(
+            {pindexPrev->GetBlockHash(), pindexPrev->hashMerkleRoot},
+            dummyNumLayers));
+        pindexPrev = pindexPrev->pprev;
```

Comment:
Do we need some check in here. I realize we have so many blocks as to not hit the genesis after this is activated, but seems prudent?